File: app/src/main/java/app/otterling/monitoring/ScreenshotUploader.kt
package app.otterling.monitoring

import android.content.Context
import android.provider.Settings
import android.util.Base64
import android.util.Log
import app.otterling.alerts.MacTamperPollSettings
import app.otterling.content.CloudFilterSettings
import java.net.HttpURLConnection
import java.net.URL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject

/** classification is one of "safe", "nsfw", "skipped" (visualFilterEnabled off server-side), or
 * "error" (classifier unavailable -- treat exactly like "safe", never block on this). */
data class ScreenshotClassifyResult(val classification: String, val blockUntilMillis: Long?)

/**
 * Uploads a captured foreground-app screenshot to filter-server's `/screenshot-classify` (see
 * `lockprofile_service.py` and `nsfw_image_classifier.py`) for server-side NSFW classification --
 * see FocusGuardAccessibilityService.kt's capture path for why this is server-side, not on-device.
 *
 * Same `LOCKPROFILE_TOKEN`/host config as [DeviceLogUploader] -- one shared device-bearer
 * credential for every low-stakes phone->server call, not a separate one per feature.
 */
object ScreenshotUploader {
    suspend fun upload(
        context: Context,
        packageName: String,
        imageBytes: ByteArray,
    ): Result<ScreenshotClassifyResult> = withContext(Dispatchers.IO) {
        runCatching {
            val settings = MacTamperPollSettings(context)
            check(settings.isConfigured()) { "Server token not configured (see Mac tamper alerts settings)" }

            val host = CloudFilterSettings(context).host()
            check(host.isNotBlank()) { "Filter server host not configured" }

            val deviceId = Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID)
                ?: "unknown-device"
            val imageBase64 = Base64.encodeToString(imageBytes, Base64.NO_WRAP)
            val requestBody = JSONObject()
                .put("device_id", deviceId)
                .put("package_name", packageName)
                .put("image_base64", imageBase64)
                .toString()

            val connection = URL("https://$host/screenshot-classify").openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "POST"
                connection.doOutput = true
                connection.connectTimeout = 15_000
                // Server-side classification can take up to CLAUDE_TIMEOUT_SECONDS=45s (vision
                // model round trip) -- longer than DeviceLogUploader's plain-upload timeout.
                connection.readTimeout = 60_000
                connection.setRequestProperty("Authorization", "Bearer ${settings.token()}")
                connection.setRequestProperty("Content-Type", "application/json")
                connection.outputStream.use { it.write(requestBody.toByteArray(Charsets.UTF_8)) }
                val code = connection.responseCode
                check(code in 200..299) { "Upload failed: HTTP $code" }
                val responseBody = connection.inputStream.bufferedReader().use { it.readText() }
                val json = JSONObject(responseBody)
                ScreenshotClassifyResult(
                    classification = json.optString("classification", "error"),
                    blockUntilMillis = if (json.isNull("blockUntilMillis")) null else json.optLong("blockUntilMillis"),
                )
            } finally {
                connection.disconnect()
            }
        }.onFailure { error ->
            Log.w(TAG, "Screenshot upload failed", error)
        }
    }

    private const val TAG = "ScreenshotUploader"
}

File: macos/FocusLock/Sources/FocusLockShared/ScreenshotUploader.swift
import Foundation

/// Uploads a captured foreground-app screenshot to filter-server's `/screenshot-classify` (see
/// `filter-server/lockprofile_service.py` and `nsfw_image_classifier.py`) for server-side NSFW
/// classification -- the exact same route and JSON shape the Android app's `ScreenshotUploader.kt`
/// already uses. `device_id`/`package_name`/`image_base64` are platform-agnostic server-side, so
/// nothing on the filter-server needed to change for a Mac to use this route too.
///
/// See `FocusLockScanner/ScreenshotMonitor.swift`'s doc comment for why capture/upload happens
/// from the per-user `FocusLockScanner` LaunchAgent rather than the root `FocusLockHelperd` daemon.
public enum ScreenshotUploader {
    public struct ClassifyResult {
        /// One of "safe", "nsfw", "skipped" (guardian disabled `visualFilterEnabled` server-side),
        /// or "error" (classifier unavailable) -- see `nsfw_image_classifier.py`. Callers must treat
        /// anything other than "nsfw" as non-blocking, matching the phone's contract exactly.
        public let classification: String
        public let blockUntilMillis: Double?
    }

    public enum UploadError: Error {
        case notConfigured
        case invalidResponse
    }

    public static func upload(
        deviceID: String,
        packageName: String,
        imageData: Data,
        completion: @escaping (Result<ClassifyResult, Error>) -> Void
    ) {
        let host = TamperReporter.resolvedHost()
        let token = TamperReporter.resolvedToken()
        guard !host.isEmpty, !token.isEmpty, let url = URL(string: "https://\(host)/screenshot-classify") else {
            completion(.failure(UploadError.notConfigured))
            return
        }

        let body: [String: Any] = [
            "device_id": deviceID,
            "package_name": packageName,
            "image_base64": imageData.base64EncodedString(),
        ]
        guard let payload = try? JSONSerialization.data(withJSONObject: body) else {
            completion(.failure(UploadError.invalidResponse))
            return
        }

        var request = URLRequest(url: url, timeoutInterval: 60)
        request.httpMethod = "POST"
        request.httpBody = payload
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error {
                completion(.failure(error))
                return
            }
            guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode),
                  let data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let classification = json["classification"] as? String else {
                completion(.failure(UploadError.invalidResponse))
                return
            }
            let blockUntilMillis = json["blockUntilMillis"] as? Double
            completion(.success(ClassifyResult(classification: classification, blockUntilMillis: blockUntilMillis)))
        }.resume()
    }
}
